src/precompiles/ecrecover.rs

```diff
@@ -0,0 +1,230 @@
+        let hash = buffer;
+
+        r_value.to_big_endian(&mut buffer[..]);
+        let r_bytes = buffer;
+
+        s_value.to_big_endian(&mut buffer[..]);
+        let s_bytes = buffer;
+
+        v_value.to_big_endian(&mut buffer[..]);
+        let v = buffer[31];
+        assert!(v == 0 || v == 1);
+
+        let pk = ecrecover_inner(&hash, &r_bytes, &s_bytes, v);
+
+        // here it may be possible to have non-recoverable k*G point, so can fail
+        if let Ok(recovered_pubkey) = pk {
+            let pk = k256::PublicKey::from(&recovered_pubkey);
+            let affine_point = *pk.as_affine();
+            use k256::elliptic_curve::sec1::ToEncodedPoint;
+            let pk_bytes = affine_point.to_encoded_point(false);
+            let pk_bytes_ref: &[u8] = pk_bytes.as_ref();
+            assert_eq!(pk_bytes_ref.len(), 65);
+            debug_assert_eq!(pk_bytes_ref[0], 0x04);
+            let address_hash = sha3::Keccak256::digest(&pk_bytes_ref[1..]);
+
+            let mut address = [0u8; 32];
+            let hash_ref: &[u8] = address_hash.as_ref();
+            address[12..].copy_from_slice(&hash_ref[12..]);
+
+            let mut write_location = MemoryLocation {
+                page: params.memory_page_to_write,
+                index: params.output_memory_offset,
+            };
+
+            let ok_marker = U256::one();
+            let ok_or_err_query = MemoryQuery {
+                location: write_location,
+                value: ok_marker,
+                value_is_pointer: false,
+                rw_flag: true,
+            };
+            heaps.execute_partial_query(ok_or_err_query)?;
+
+            write_location.index += 1;
+            let result = U256::from_big_endian(&address);
+            let result_query = MemoryQuery {
+                location: write_location,
+                value: result,
+                value_is_pointer: false,
+                rw_flag: true,
+            };
+            heaps.execute_partial_query(result_query)?;
+        } else {
+            let mut write_location = MemoryLocation {
+                page: params.memory_page_to_write,
+                index: params.output_memory_offset,
+            };
+
+            let err_marker = U256::zero();
+            let ok_or_err_query = MemoryQuery {
+                location: write_location,
+                value: err_marker,
+                value_is_pointer: false,
+                rw_flag: true,
+            };
+            heaps.execute_partial_query(ok_or_err_query)?;
+
+            write_location.index += 1;
+            let empty_result = U256::zero();
+            let result_query = MemoryQuery {
+                location: write_location,
+                value: empty_result,
+                value_is_pointer: false,
+                rw_flag: true,
+            };
+            heaps.execute_partial_query(result_query)?;
+        }
+        Ok(())
+    }
+}
+
+pub fn ecrecover_inner(
```

Comment:
Same as above, let's try and add a comment -- however small -- on what we think this function does. Seems like this is Elliptic Curve Recover of a pair, but feel free to correct me.